raft topology: replace publish_cdc_generation with a bg fiber

patjed41 opened the pull request:
Currently, the topology coordinator has the `topology::transition_state::publish_cdc_generation` state responsible for publishing the already created CDC generations to the user-facing description tables. This process cannot fail as it would cause some CDC updates to be missed. On the other hand, we would like to abort the `publish_cdc_generation` state when bootstrap aborts. Of course, we could also wait until handling this state finishes, even in the case of the bootstrap abort, but that would be inefficient. We don't want to unnecessarily block topology operations by publishing CDC generations.

The solution proposed by this PR is to remove the `publish_cdc_generation` state completely and introduce a new background fiber of the topology coordinator -- `cdc_generation_publisher` -- that continually publishes committed CDC generations.

Apart from introducing the CDC generation publisher, we add `test_cdc_generation_publishing.py` that verifies its correctness and we adapt other CDC tests to the new changes.

Fixes #15194

patjed41 commented:
In #15194, the proposed solution was to modify `system_distributed_keyspace::create_cdc_desc` and publish all CDC generations atomically at once. I don't know if it is worth implementing. For now, I've implemented an easier iterative solution. Although it seems less efficient, the number of unpublished CDC generations shouldn't exceed one in vast majority of cases, so in practice there should be almost no difference. Changing the current solution to the proposed one requires only a few small modifications of the current PR (apart from `system_distributed_keyspace::create_cdc_desc`) so I can easily change it if we decide to go for it.

gleb-cloudius commented:
> In #15194, the proposed solution was to modify `system_distributed_keyspace::create_cdc_desc` and publish all CDC generations atomically at once. I don't know if it is worth implementing. For now, I've implemented an easier iterative solution. Although it seems less efficient, the number of unpublished CDC generations shouldn't exceed one in vast majority of cases, so in practice there should be almost no difference. Changing the current solution to the proposed one requires only a few small modifications of the current PR (apart from `system_distributed_keyspace::create_cdc_desc`) so I can easily change it if we decide to go for it.

@kbr-scylla asked for atomicity since otherwise a driver may see later generations timestamps earlier and miss some updates. By "iterative" solution do you mean that you publish CDC generations one by one and do not start later one until earlier is fully published?

patjed41 commented:
> By "iterative" solution do you mean that you publish CDC generations one by one and do not start later one until earlier is fully published?

The generations are published in non-decreasing order of timestamps.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -1191,6 +1197,69 @@ class topology_coordinator {
+                }
+            } catch (raft::request_aborted&) {
+                slogger.debug("raft topology: CDC generation publisher fiber aborted");
+            } catch (term_changed_error&) {
```
I do not think we should ever get it here. Happens when we try to retake a node.

gleb-cloudius replied:
No, I am taking this back. start_operation() may throw it. _as should be aborted at this point, but we do not know what happens first us checking _as or getting the exception.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -410,8 +410,6 @@ future<> storage_service::topology_state_load() {
                     [[fallthrough]];
                 case topology::transition_state::commit_cdc_generation:
                     [[fallthrough]];
-                case topology::transition_state::publish_cdc_generation:
```
`update_topology_state` does not handle `group0_concurrent_modification`. It only logs and re-throws.

patjed41 replied:
Right, I somehow missed this `throw;`. I fixed it.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3473/

patjed41 commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3473/

Waiting for the message in the `cdc_generation_publisher_fiber` error injection added by me timed out in debug mode. Logs suggest to me that the 1 minute timout is almost enough. For now, I increased the timeout from 1 minute to 2 minutes and I think it should be enough to make `test_multiple_unpublished_cdc_generations` pass every time, but we will see. Another option is to skip this test in debug mode.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -1227,6 +1227,12 @@ class topology_coordinator {
         while (!_as.abort_requested()) {
+            co_await utils::get_local_injector().inject_with_handler("cdc_generation_publisher_fiber", [] (auto& handler) -> future<> {
+                slogger.info("raft toplogy: CDC generation publisher fiber sleeps after injection");
+                co_await handler.wait_for_message(std::chrono::steady_clock::now() + std::chrono::minutes{2});
```
Very optimistic to assume that you'll manage to bootstrap 3 nodes in 2 minutes on our great CI machines with all these concurrent tests :) And let's not even talk about ARM.

Let's increase the timeout to 5 minutes. And... let's mark the two tests dev-mode only.

patjed41 replied:
Done.

kbr-scylla commented on test/topology_experimental_raft/test_cdc_generation_publishing.py:
```diff
@@ -0,0 +1,106 @@
+        await handler.message()
+        while len(gen_timestamps) < 4:
+            # We prefer to detect CDC generation publications one-by-one, because it increases our chances of catching
+            # potential bugs like incorrect order of publications. Therefore, we use very short period - 0.01 s.
```
Nice.

kbr-scylla commented on test/cql-pytest/test_cdc.py:
```diff
@@ -7,12 +7,24 @@
+    query = SimpleStatement(
+            "select time from system_distributed.cdc_generation_timestamps where key = 'timestamps'",
+            consistency_level=ConsistencyLevel.ONE)
+    while len(list(cql.execute(query))) == 0:
```
This should have a time limit.

patjed41 replied:
Done.

patjed41 commented on service/topology_state_machine.hh:
```diff
@@ -137,6 +137,9 @@ struct topology {
     // It's used as partition key in CDC_GENERATIONS_V3 table.
     std::optional<utils::UUID> new_cdc_generation_data_uuid;
 
+    // The IDs of the commited yet unpublished CDC generations sorted by timestamps.
```
Yes, exactly.

patjed41 commented:
v3:
- fixed the name of `set_unpublisheded_cdc_generations`
- increased the timeout of waiting for the message in the `cdc_generation_publisher_fiber` error injection to 5 minutes
- made `test_cdc_generation_publishing.py` dev-only
- added timeout to `wait_for_first_cdc_generation`

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3530/

patjed41 commented:
I only rebased on newer `master`. I realized #15319 conflicts with this PR and I need to rebase it sooner or later, but I must firstly match their versions of `master`.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3546/
